### DISCUSSION TITLE
Assemblies/Multi-file

### DISCUSSION THREAD

**Discussion by jgomez720** [KittyCAD/modeling-app#494] [New Feature Requests]
### Description
User can combine multiple files into one assembly file and perform mating features such as touch, align, concentric, parallel, center, etc.

### Functionality
Some of this is up to our choosing. We can either have the user create a designated assembly file or just have them add another file to their existing file and create the assembly from there. Or maybe we don't even create a separate file. SolidWorks has separate files for piece parts and assemblies (.sldprt and .sldasm). NX doesn't (.prt for both piece parts and assemblies).

### Supporting Docs/Images
https://github.com/KittyCAD/modeling-app/assets/114548659/17601807-157d-4b94-b047-06df4d153031

**Comment by franknoirot** [comment]
My first questions for the group:
1. Should assemblies be a file unto themselves, or a collection of features of KCL that allow any KCL file to be an assembly? My instinct is the latter because we already have an `import` function, and we can encourage assemblies as their own files just for organizational purposes by convention.
2. How will we allow users/AST mods to reach in and tag and reference various faces/edges/points within imported KCL geometry?
3. How will we allow the same for non-KCL geometry?

**Comment by jgomez720** [comment]
I really only have an answer to 1. I think whatever allows for easier automatic BOM generation down the line. I think it would be easy to scan the imports and then see how many times it was called in the KCL file.
```
import fastener

(example KCL)

fastener([-2, -2])
fastener([2, 2])
fastener([2, -2])
fastener([-2, 2])
```
This way, we know they used 4x fasteners in the assembly. So my vote is also the latter

**Comment by franknoirot** [comment]
> 2. How will we allow users/AST mods to reach in and tag and reference various faces/edges/points within imported KCL geometry?

This may be on its way to being solved thanks to scoped tags. I think we still need to unlock the ability for these tags to "roll up" in the returned geometry, so an ExtrudeGroup or RevolveGroup inherits the tags of its source sketch too, but it's close.

**Reply by franknoirot** [reply]
Aged like milk, other team members should chime in about how the `.tags` thing makes for a fragile public API for any imported geometry
